### PR TITLE
[FLINK-14212][python] Support no-argument Python UDFs.

### PR DESCRIPTION
**Pull request by WeiZhong94**
## What is the purpose of the change

*This pull request fixes the problem that no-argument Python UDFs cannot work in Flink Python API.*

## Brief change log

  - *Filter the RexNodes that contains PythonFunction when optimizing const expressions in ExpressionReducer.*

## Verifying this change

This change is already covered by existing tests, such as *ExpressionReducerTest* and *test_udf.py*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-python/pyflink/table/tests/test_udf.py [hunk 1]
```diff
@@ -204,6 +204,26 @@ def eval(self, col):
...
+        self.t_env.register_function(
+            "one", one)
+        self.t_env.register_function("add", add)
+
+        table_sink = source_sink_utils.TestAppendSink(['a', 'b'],
+                                                      [DataTypes.BIGINT(), DataTypes.BIGINT()])
+        self.t_env.register_table_sink("Results", table_sink)
+
+        t = self.t_env.from_elements([(1, 2), (2, 5), (3, 1)], ['a', 'b'])
+        t.select("one(), add(a, b)") \
```
[the remark this answers is not in the document]
Comment (WeiZhong94) [reply flink-python/pyflink/table/tests/test_udf.py]: My initial thought was to use the "add" function to test whether the no-argument UDF can work with normal UDFs. It seems such worries are redundant so I have remove it in the latest commit.

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your comments! I have addressed them in the latest commit, please take a look.

File: flink-python/pyflink/table/tests/test_udf.py [hunk 2]
```diff
@@ -204,6 +204,28 @@ def eval(self, col):
             self.t_env.register_function(
                 "non-callable-udf", udf(Plus(), DataTypes.BIGINT(), DataTypes.BIGINT()))
 
+    def test_udf_without_arguments(self):
+        @udf(input_types=[], result_type=DataTypes.BIGINT(), deterministic=True)
```
Comment (dianfu) [inline flink-python/pyflink/table/tests/test_udf.py]: 1. If there is no arguments, is it possible not declare input_types?
2. What about using lambda expression instead? Then just one line for each udf

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExpressionReducer.scala [hunk 3]
```diff
@@ -24,20 +24,20 @@ import org.apache.flink.metrics.MetricGroup
 import org.apache.flink.table.api.{TableConfig, TableException}
 import org.apache.flink.table.dataformat.BinaryStringUtil.safeToString
 import org.apache.flink.table.dataformat.{BinaryString, Decimal, GenericRow}
-import org.apache.flink.table.functions.{FunctionContext, UserDefinedFunction}
+import org.apache.flink.table.functions.{FunctionContext, FunctionLanguage, UserDefinedFunction}
 import org.apache.flink.table.planner.calcite.FlinkTypeFactory
 import org.apache.flink.table.planner.codegen.FunctionCodeGenerator.generateFunction
 import org.apache.flink.table.runtime.functions.SqlDateTimeUtils
 import org.apache.flink.table.types.logical.RowType
-
```
Comment (dianfu) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExpressionReducer.scala]: unnecessary change

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExpressionReducer.scala [hunk 4]
```diff
@@ -24,20 +24,20 @@ import org.apache.flink.metrics.MetricGroup
...
 import org.apache.calcite.avatica.util.ByteString
 import org.apache.calcite.rex.{RexBuilder, RexExecutor, RexNode}
 import org.apache.calcite.sql.`type`.SqlTypeName
 import org.apache.commons.lang3.StringEscapeUtils
-
```
Comment (dianfu) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExpressionReducer.scala]: unnecessary change

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExpressionReducer.scala [hunk 5]
```diff
@@ -24,20 +24,20 @@ import org.apache.flink.metrics.MetricGroup
...
 import java.io.File
 import java.util.TimeZone
 
+import org.apache.flink.table.planner.plan.utils.PythonUtil.FunctionFinder
```
Comment (dianfu) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExpressionReducer.scala]: Correct the order

**Comment by WeiZhong94** [comment]
@JingsongLi I noticed that the "getScalarFunction" method in "ScalarSqlFunction" of blink planner was replaced by "makeFunction" method. I think using "getScalarFunction" to get the initial ScalarFunction object and check its language type makes more sense here so I re-add the method. Please take a look at this PR to make sure it does not cause any side effects to blink planner :).

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/PythonUtil.scala [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import scala.collection.JavaConversions._
+
+object PythonUtil {
+  /**
+    * Checks whether it contains the specified kind of function in a RexNode.
+    *
+    * @param expectedLanguage the expected kind of function to find
+    * @param recursive whether check the inputs
+    */
+  class FunctionFinder(expectedLanguage: FunctionLanguage, recursive: Boolean)
```
Comment (dianfu) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/PythonUtil.scala]: There is a method `containsFunctionOf`, could you also port it to blink planner and also make `FunctionFinder` private?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/ExpressionReductionRulesTest.scala [hunk 7]
```diff
@@ -20,9 +20,9 @@ package org.apache.flink.table.planner.plan.rules.logical
 
 import org.apache.flink.api.scala._
 import org.apache.flink.table.api.scala._
+import org.apache.flink.table.functions.{FunctionLanguage, ScalarFunction}
 import org.apache.flink.table.planner.expressions.utils.{Func1, RichFunc1}
 import org.apache.flink.table.planner.utils.TableTestBase
-
```
Comment (dianfu) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/ExpressionReductionRulesTest.scala]: unnecessary change

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/ExpressionReductionRulesTest.scala [hunk 8]
```diff
@@ -46,4 +46,21 @@ class ExpressionReductionRulesTest extends TableTestBase {
...
 
+  @Test
+  def testExpressionReductionWithPythonUDF(): Unit = {
+    util.addFunction("PyUdf", MockedPythonScalarFunction)
+    util.addFunction("MyUdf", Func1)
+    util.verifyPlan("SELECT PyUdf(), MyUdf(1) FROM MyTable")
+  }
+}
+
+object MockedPythonScalarFunction extends ScalarFunction {
```
Comment (dianfu) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/ExpressionReductionRulesTest.scala]: rename to PythonUDFWithoutArgument?

File: flink-python/pyflink/table/tests/test_udf.py (hunk 2 above)
Comment (WeiZhong94) [reply flink-python/pyflink/table/tests/test_udf.py]: 1. In current design it will throw an exception if we don't declare input_types. And IMO the lack of input_types should not be considered as a no-argument UDF. It means we should infer the input types when the UDF is called. 
2. I have replaced current UDFs with lambda in the latest commit.

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your review again! I have addressed your comments in the latest commit.

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your review! I have addressed the comment in the latest commit.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExpressionReducer.scala [hunk 9]
```diff
@@ -120,59 +124,64 @@ class ExpressionReducer(
...
-          val value = if (reducedValue != null &&
-            unreduced.getType.getSqlTypeName == SqlTypeName.DOUBLE) {
-            new java.math.BigDecimal(reducedValue.asInstanceOf[Number].doubleValue())
-          } else {
-            reducedValue
-          }
-
-          reducedValues.add(maySkipNullLiteralReduce(rexBuilder, value, unreduced))
-          reducedIdx += 1
+      if (PythonUtil.containsFunctionOf(unreduced, FunctionLanguage.PYTHON)) {
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/ExpressionReducer.scala]: Rex tree traverse is expensive. We traverse twice in this reduce() method. How about do some reuse here, e.g., 
```
    val jvmUdfs = constExprs.asScala.zipWithIndex
      .filter(p => PythonUtil.containsFunctionOf(p._1, FunctionLanguage.JVM))
...

     if (!jvmUdfIndexes.contains(i)) {
        // if contains python function then just insert the original expression.
        reducedValues.add(unreduced)
      }
```
In this way, we can reduce some optimization time. What do you think?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/ExpressionReductionRulesTest.scala [hunk 10]
```diff
@@ -46,4 +47,21 @@ class ExpressionReductionRulesTest extends TableTestBase {
...
 
+  @Test
+  def testExpressionReductionWithPythonUDF(): Unit = {
+    util.addFunction("PyUdf", MockedPythonUDFWithoutArguments)
+    util.addFunction("MyUdf", Func1)
+    util.verifyPlan("SELECT PyUdf(), MyUdf(1) FROM MyTable")
+  }
+}
+
+object MockedPythonUDFWithoutArguments extends ScalarFunction {
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/ExpressionReductionRulesTest.scala]: Check both deterministic and non-deterministic python udfs?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/ExpressionReductionRulesTest.scala [hunk 11]
```diff
@@ -510,3 +528,14 @@ object DeterministicNullFunc extends ScalarFunction {
   def eval(): String = null
   override def isDeterministic = true
 }
+
+object MockedPythonUDFWithoutArguments extends ScalarFunction {
+
+  override def getLanguage: FunctionLanguage = FunctionLanguage.PYTHON
+
+  def eval(): Long = {
```
Comment (hequn8128) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/ExpressionReductionRulesTest.scala]: Simplify the eval to `def eval(): Long = 1L` directly? and change the class name to `DeterministicPythonFunc`.

**Comment by WeiZhong94** [comment]
@hequn8128 Thanks for your feedback! Dian Fu and me has talked about these two approach and we come to an agreement that skip the optimization in `ExpressionReducer` and optimize the UDFs during runtime. Our thoughts are as follows:
1. How to optimize Python UDFs during runtime?
After support constant parameters in Python UDF(see [this PR](https://github.com/apache/flink/pull/9858)), we can do this optimization when evaluating the chained Python UDFs in python worker: 
If the evaluated Python UDF is deterministic and has no argument or its arguments are all constant value, which means it will always return a constant value, replace it with the constant result value.
This rule can be applied recursively until all the deterministic UDFs with constant inputs have been replaced. If the root nodes of the chained Python UDFs become constant values, we can only transmit them only once and replace them with None in following transmission to save IO resource. The Java operator also knows which fields of the evaluated result should be constant value rather than None because the reduce rule can be applied in Java side too. No additional interaction between Java operator and Python worker is needed in this design.
2. Why not optimize Python UDFs during optimization phase?
Reducing Python UDFs in optimization phase is not a easy work in current design. It means the generated Java wrappers of Python UDFs can be evaluated and return the correct result. In other word we need run Python UDFs in client side, but the Python UDFs is designed to run on cluster whose python environment may different from client side after we introduce environment and dependency management in the future. To solve the environment problem, we need to prepare a python environment that is the same as the python environment on cluster before reducing Python UDFs. To evaluate the Python UDF, we need to implement a new UDF runner which does not depend on Apache Beam(the client side of Flink Python API does not depend on Apache Beam). We know if we complete these it will be a perfect solution of this problem, but it is too expensive compared to runtime optimization.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/ExpressionReductionRulesTest.scala (hunk 10 above)
Comment (WeiZhong94) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/ExpressionReductionRulesTest.scala]: IMO it's not necessary to check non-deterministic python UDFs in this rule test. My reason is:
1. If the UDF is non-deterministic, it won't match this rule whether it is a Python UDF or not. This test case won't cover more code.
2. We already have a ITCase about non-deterministic python UDF in `test_udf.py` to make sure the non-deterministic python UDFs work properly.

**Comment by WeiZhong94** [comment]
@hequn8128 Thanks for your review! I have addressed your comments in the latest commit.

**Comment by JingsongLi** [comment]
> @JingsongLi I noticed that the "getScalarFunction" method in "ScalarSqlFunction" of blink planner was replaced by "makeFunction" method. I think using "getScalarFunction" to get the initial ScalarFunction object and check its language type makes more sense here so I re-add the method. Please take a look at this PR to make sure it does not cause any side effects to blink planner :).

LGTM, it is reasonable, can you just use scala val to scalarFunction just like `TableSqlFunction`?

**Comment by WeiZhong94** [comment]
> > @JingsongLi I noticed that the "getScalarFunction" method in "ScalarSqlFunction" of blink planner was replaced by "makeFunction" method. I think using "getScalarFunction" to get the initial ScalarFunction object and check its language type makes more sense here so I re-add the method. Please take a look at this PR to make sure it does not cause any side effects to blink planner :).
> 
> LGTM, it is reasonable, can you just use scala val to scalarFunction just like `TableSqlFunction`?

Thanks for your reply! That makes sense to me. I have updated the code in the latest commit.

**Comment by hequn8128** [comment]
@WeiZhong94 Thanks a lot for the update. Also thank you all for the review.
LGTM, will merge this once travis passed.
